File: website/kohomology-js/src/main/kotlin/styled.kt
@ExperimentalJsExport
@JsExport
@Suppress("UNUSED")
class StyledStringKt(
    val stringType: String,
    val content: String,
)

enum class StringType(val typeName: String) {
    TEXT("text"),
    MATH("math"),
}

class StyledStringInternal(
    private val stringType: StringType,
    private val content: String,
) {
    @ExperimentalJsExport
    fun export(): StyledStringKt {
        return StyledStringKt(this.stringType.typeName, this.content)
    }
}

@ExperimentalJsExport
@JsExport
@Suppress("UNUSED")
class StyledMessageKt(
    val messageType: String,
    val strings: Array<StyledStringKt>,
)

enum class MessageType(val typeName: String) {
    SUCCESS("success"),
    ERROR("error"),
}

class StyledMessageInternal(
    val messageType: MessageType,
    val strings: List<StyledStringInternal>,
) {
    @ExperimentalJsExport
    fun export(): StyledMessageKt {
        val strings = this.strings.map { it.export() }.toTypedArray()
        return StyledMessageKt(this.messageType.typeName, strings)
    }
}

object StyledStringListBuilder {
    val String.text: List<StyledStringInternal>
        get() = listOf(StyledStringInternal(StringType.TEXT, this))
    val String.math: List<StyledStringInternal>
        get() = listOf(StyledStringInternal(StringType.MATH, this))
}

fun styledMessage(messageType: MessageType, block: StyledStringListBuilder.() -> List<StyledStringInternal>): StyledMessageInternal {
    return StyledMessageInternal(messageType, StyledStringListBuilder.block())
}

File: website/src/components/Calculator/styled.ts
import { StyledStringKt, StyledMessageKt } from "kohomology-js"

const stringTypes = ["normal", "math"] as const
type StringType = typeof stringTypes[number]

export interface StyledString {
  readonly stringType: StringType
  readonly content: string
}

const messageTypes = ["success", "error"] as const
type MessageType = typeof messageTypes[number]

export interface StyledMessage {
  readonly messageType: MessageType
  readonly strings: StyledString[]
}

export function fromString(messageType: MessageType, str: string): StyledMessage {
  const styledString: StyledString = {
    stringType: "normal",
    content: str,
  }
  return {
    messageType: messageType,
    strings: [styledString]
  }
}


export function toStyledString(styledStringKt: StyledStringKt): StyledString {
  const stringType: string = styledStringKt.stringType
  if (!(stringTypes as readonly string[]).includes(stringType)) {
    throw new Error(`Invalid stringType: ${stringType}`)
  }
  return {
    stringType: stringType as StringType,
    content: styledStringKt.content
  }
}

export function toStyledMessage(styledMessageKt: StyledMessageKt): StyledMessage {
  const messageType: string = styledMessageKt.messageType
  if (!(messageTypes as readonly string[]).includes(messageType)) {
    throw new Error(`Invalid messageType: ${messageType}`)
  }
  return {
    messageType: messageType as MessageType,
    strings: styledMessageKt.strings.map(toStyledString)
  }
}
